### DISCUSSION TITLE
Chat message queue with configurable delay

### DISCUSSION THREAD

**Discussion by lewisgibson** [Chatterino/chatterino2#3549] [Ideas]
Twitch's native moderation product has a feature to delay messages being rendered for viewers that are not moderators or the broadcaster. This is especially useful for large channels that receive a lot of hate as it allows bad messages to be filtered before being visible to the public. I would really like to see this feature ported to Chatterino so the automated moderation infrastructure I build to support streamers can act without the original bad messages being visible on their chat window during the live stream. 

Although the native product only allows increments of one second, I would be looking for ~500ms granularity at a minimum if possible. This allows _most_ automated actions to occur without introducing a noticeable delay to their chat.

**Comment by Felanbird** [comment]
Originally discussed all the way back in #645 - but the implementation for chromecast/tv streams was pretty megadank, but a very small and unnoticeable delay seems very reasonable.

**Comment by pajlada** [comment]
This is how I imagine the options for this feature to look like:

Twitch chat delay:
    A dropdown with the following options:
        Use Twitch Delay (whatever we get from the Twitch API). NOTE: Can this be seen anywhere?
        Disabled/Off
        Custom delay
            If custom delay is chosen, show a QDoubleSpinBox with steps of 0.5
    Default Value: Use Twitch Delay (Default behaviour on Twitch) or Disabled (Default behaviour in Chatterino) - pros/cons appreciated
    Store all these values as an integer in the settings. -1 meaning Use Twitch Delay, 0 meaning Disabled/Off, anything else meaning how many milliseconds to client-side delay.

Enable chat delay as broadcaster/mod: [ ]
    Default Value: Off
    Description/tooltip: By default, the Twitch chat delay option only applies to channels where you are not a moderator or broadcaster

**Comment by dnsge** [comment]
I decided to do a little bit of research into this feature. In terms of fetching the chat delay setting, I found a couple of things:
1. There is the [Get Chat Settings](https://dev.twitch.tv/docs/api/reference#get-chat-settings) API endpoint. However, getting the `non_moderator_chat_delay_duration` value requires (for reasons I don't understand?) the `moderator:read:chat_settings` scope.

2. There is a GQL request that the frontend actually uses, shown below [1], but as we know we can't use GQL.

So at the moment, it seems like there isn't any way to fetch and use the chat delay value automatically. We'd only be able to support users willingly delaying their chat (?)

-------
[1] GQL Request:
```json5
{
    "operationName": "MessageBuffer_Channel",
    "variables": {
        "channelLogin": "<channel login>"
    },
    "extensions": {
        "persistedQuery": {
            "version": 1,
            "sha256Hash": "bfc959904f55b5003ae4674d4bea83ebdcd8867ad76e12f38957d433902d2fcc"
        }
    }
}

// Response
{
    "data": {
        "user": {
            "id": "<channel id>",
            "chatSettings": {
                // 0, 2000, 4000, or 6000 ms based on moderation settings
                "chatDelayMs": 6000, 
                "__typename": "ChatSettings"
            },
            "__typename": "User"
        }
    },
    "extensions": {
        "durationMilliseconds": 23,
        "operationName": "MessageBuffer_Channel",
        "requestID": "XXX"
    }
}
```

**Reply by Felanbird** [reply]
> We'd only be able to support users willingly delaying their chat

Yeah that's the only approach we'd be able to take.

> However, getting the non_moderator_chat_delay_duration value requires (for reasons I don't understand?) the moderator:read:chat_settings scope.

It's like that since chat delay isn't publicly exposed, (only accessible so mods can know what it is, and set it) a lot of users are unaware about how the delay works until they switch to Chatterino.

**Reply by dnsge** [reply]
Well, I suppose it's not publicly exposed in the sense that there isn't a "show chat delay" button that a regular user can press, but obviously every user, even anonymous, has access to the value so that the delay can be imposed.

It follows the recurring theme of Twitch making GQL's publically-accessible data inaccessible through Helix (see: polls, predictions)
